src/highlight.js

```diff
@@ -365,6 +365,23 @@ const HLJS = function(hljs) {
         }
       }
 
+      // edge case for when illegal matches $ (end of line) which is technically
+      // a 0 width match but not a begin/end match so it's not caught by the
+      // first handler (when ignoreIllegals is true)
+      if (match.type === "illegal" && lexeme === "") {
+        // advance so we aren't stuck in an infinite loop
+        return 1;
+      }
+
+      // infinite loops are BAD, this is a last ditch catch all. if we have a
+      // decent number of iterations yet our index (cursor position in our
+      // parsing) still 3x behind our index then something is very wrong
+      // so we bail
+      if (iterations > 100000 && iterations > match.index * 3) {
+        const err = new Error('potential infinite loop, way more iterations than matches');
+        throw err;
+      }
```

Comment:
Sorry I'm late to the party. Quick fix was the right thing to do.
I'd move this check out to the `for (;;)` loop, in order not to deal with `iterations` inside `processLexeme` function.

Reply:
Except then we'd literally have to check it ALL the time... where-as right now it's more of a fallback... but I'll play around with it.

Reply:
But technically your right in that if there were a bug in the above code that this wouldn't be well suited to catch it... :-/